scene/gui/button.cpp

```diff
@@ -309,16 +310,28 @@ void Button::set_icon(const Ref<Texture> &p_icon) {
 	if (icon == p_icon) {
 		return;
 	}
+	if (icon.is_valid()) {
+		icon->disconnect(SceneStringNames::get_singleton()->changed, this, "_texture_changed");
+	}
 	icon = p_icon;
-	update();
 	_change_notify("icon");
-	minimum_size_changed();
+	if (icon.is_valid()) {
+		icon->connect(SceneStringNames::get_singleton()->changed, this, "_texture_changed");
+	}
+	_texture_changed();
 }
 
 Ref<Texture> Button::get_icon() const {
 	return icon;
 }
 
+void Button::_texture_changed() {
+	if (icon.is_valid()) {
+		update();
+		minimum_size_changed();
+	}
+}
```

Comment:
So one note here is that this is now what's called in `set_icon`, whereas in the 4.x version we just call the same methods unconditionally. It's a bit of a strange change between the two versions, but also I'm not sure why the check is here to begin with, in both 3.x and 4.x versions. In `TextureButton` we don't add a condition and it makes sense that we don't.

Reply:
The 4.x version does the check, taken directly from there

Reply:
To add to this, the main risk I envision is no update happen when going from valid icon to no icon. (Currently compiling to check if it's indeed a problem.)

Reply:
Restored the order in the `set_icon` method

Reply:
I don't really recall why I added that check in 4.x, can remove it here and in 4.x too

Reply:
I think the check is redundant because currently the only caller is the connected `changed` signal above.

Reply:
Agreed, the one in `TextureRect` looks like a leftover from earlier versions, leaving it in for the moment for parity